server/schedule/checker_controller.go

```diff
@@ -34,64 +38,80 @@
 	ruleChecker       *checker.RuleChecker
 	mergeChecker      *checker.MergeChecker
 	jointStateChecker *checker.JointStateChecker
+	regionWaitingList cache.Cache
 }
 
 // NewCheckerController create a new CheckerController.
 // TODO: isSupportMerge should be removed.
 func NewCheckerController(ctx context.Context, cluster opt.Cluster, ruleManager *placement.RuleManager, opController *OperatorController) *CheckerController {
+	regionWaitingList := cache.NewDefaultCache(DefaultCacheSize)
 	return &CheckerController{
 		cluster:           cluster,
 		opts:              cluster.GetOpts(),
 		opController:      opController,
 		learnerChecker:    checker.NewLearnerChecker(cluster),
-		replicaChecker:    checker.NewReplicaChecker(cluster),
-		ruleChecker:       checker.NewRuleChecker(cluster, ruleManager),
+		replicaChecker:    checker.NewReplicaChecker(cluster, regionWaitingList),
+		ruleChecker:       checker.NewRuleChecker(cluster, ruleManager, regionWaitingList),
 		mergeChecker:      checker.NewMergeChecker(ctx, cluster),
 		jointStateChecker: checker.NewJointStateChecker(cluster),
+		regionWaitingList: regionWaitingList,
 	}
 }
 
 // CheckRegion will check the region and add a new operator if needed.
-func (c *CheckerController) CheckRegion(region *core.RegionInfo) (bool, []*operator.Operator) { //return checkerIsBusy,ops
+func (c *CheckerController) CheckRegion(region *core.RegionInfo) []*operator.Operator { //return checkerIsBusy,ops
 	// If PD has restarted, it need to check learners added before and promote them.
 	// Don't check isRaftLearnerEnabled cause it maybe disable learner feature but there are still some learners to promote.
 	opController := c.opController
-	checkerIsBusy := true
 
 	if op := c.jointStateChecker.Check(region); op != nil {
-		return false, []*operator.Operator{op}
+		return []*operator.Operator{op}
 	}
 
 	if c.opts.IsPlacementRulesEnabled() {
-		if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
-			checkerIsBusy = false
-			if op := c.ruleChecker.Check(region); op != nil {
-				return checkerIsBusy, []*operator.Operator{op}
+		if op := c.ruleChecker.Check(region); op != nil {
+			if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
+				return []*operator.Operator{op}
 			}
+			c.regionWaitingList.Put(region.GetID(), nil)
 		}
 	} else {
 		if op := c.learnerChecker.Check(region); op != nil {
-			return false, []*operator.Operator{op}
+			return []*operator.Operator{op}
 		}
-		if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
-			checkerIsBusy = false
-			if op := c.replicaChecker.Check(region); op != nil {
-				return checkerIsBusy, []*operator.Operator{op}
+		if op := c.replicaChecker.Check(region); op != nil {
+			if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
+				return []*operator.Operator{op}
 			}
+			c.regionWaitingList.Put(region.GetID(), nil)
```

Comment:
It seems that it can be put inside the checker or outside the checker. Can you explain the design?